presto-cli/src/main/java/com/facebook/presto/cli/QueryPreprocessor.java

```diff
@@ -0,0 +1,162 @@
+            throws QueryPreprocessorException
+    {
+        Thread clientThread = Thread.currentThread();
+        SignalHandler oldHandler = Signal.handle(SIGINT, signal -> clientThread.interrupt());
+        try {
+            if (REAL_TERMINAL) {
+                System.out.print(PREPROCESSING_QUERY_MESSAGE);
+                System.out.flush();
+            }
+            return preprocessQueryInternal(catalog, schema, query, preprocessorCommand, timeout);
+        }
+        finally {
+            if (REAL_TERMINAL) {
+                System.out.print("\r" + Strings.repeat(" ", PREPROCESSING_QUERY_MESSAGE.length()) + "\r");
+                System.out.flush();
+            }
+            Signal.handle(SIGINT, oldHandler);
+            Thread.interrupted(); // clear interrupt status
+        }
+    }
+
+    private static String preprocessQueryInternal(Optional<String> catalog, Optional<String> schema, String query, List<String> preprocessorCommand, Duration timeout)
+            throws QueryPreprocessorException
+    {
+        // execute the process in a child thread so we can better handle interruption and timeouts
+        AtomicReference<Process> processReference = new AtomicReference<>();
+        SettableFuture<String> resultFuture = SettableFuture.create();
+        Thread thread = new Thread(() -> {
+            try {
+                ProcessBuilder processBuilder = new ProcessBuilder(preprocessorCommand);
+                processBuilder.redirectErrorStream(true);
+                processBuilder.environment().put(ENV_PRESTO_CATALOG, catalog.orElse(""));
+                processBuilder.environment().put(ENV_PRESTO_SCHEMA, schema.orElse(""));
+
+                Process process = processBuilder.start();
+                try {
+                    // write query to process standard out
+                    OutputStream outputStream = process.getOutputStream();
+                    outputStream.write(query.getBytes(UTF_8));
+                    process.getOutputStream().flush();
+                    process.getOutputStream().close();
+
+                    // read response
+                    String result = CharStreams.toString(new InputStreamReader(process.getInputStream(), UTF_8));
+
+                    // wait for process to finish
+                    int exitCode = process.waitFor();
+
+                    // check we got a valid exit code
+                    if (exitCode == 0) {
+                        resultFuture.set(result);
+                    }
+                    else {
+                        resultFuture.setException(new QueryPreprocessorException("Query preprocessor failed:\n\n" + result));
+                    }
+                }
+                finally {
+                    process.destroyForcibly();
+                }
+            }
+            catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
```

Comment:
This is a pattern we use through out Presto.  When ever we catch and `InterruptedException`, we must either set the interrupted flag or throw an InterruptedException.  We could not follow that pattern here, but then we would need a comment explaining why it was allowed.  In this case, I think it is easier to follow the pattern.